### ISSUE TITLE
Sentence-level embeddings to view the distribution of topics within a single document

### ISSUE THREAD

**Issue by pankratz-l** [MaartenGr/BERTopic#362]
Hi Maarten,

I'm getting back to you regarding a question about sentence-level and document-level embeddings. I have the feeling that I have not understood something fundamental and I can't find an answer in the documentation. In your Medium article you write:

"You can use the method described in this article (or simply use BERTopic) to also create sentence-level embeddings. The main advantage of this is the possibility to view the distribution of topics within a single document."

Is there an practical guidance for this procedure?

Currently, I create the embeddings as follows:

`from sentence_transformers import SentenceTransformer`
`model = SentenceTransformer('paraphrase-multilingual-mpnet-base-v2')`
`embeddings = model.encode(doc, show_progress_bar=True)`

Each document contains several sentences and paragraphs. So I create embeddings on document-level and I can't see the probability of different topics in one document (like in LDA) or am I wrong?
Do I need to split each document into individual sentences so that each doc is one sentence?
How would I then reassign the individual sentences to the document? Is there already a function for this in BERTopic?

Thank you Maarten for your constant support!
Luca

**Comment by MaartenGr** [comment]
Hi Luca,

You indeed need to split each document up into individual sentences or paragraphs (depending on your preference/use case) and feed that to BERTopic instead of the entire documents. Then, you can assign each sentence to a document yourself by mapping the IDs of each sentence to a document. As of right now, there currently is no function in BERTopic for doing this as a sentence- or paragraph splitter would need to be implemented by the user. 

The mapping should be rather straightforward. Give each document an ID, then split those documents up into sentences and give those sentences the same ID. Then, you can count how often certain topics are mentioned in each document.

**Comment by pankratz-l** [comment]
Hi Maarten, thank you for the clarification. I am a bit confused by the term Sentence Transformers and I'm still not clear why we apply them on document level. I thought Sentence-Transformers automatically create an embedding for each sentence (separated by a dot)?

**Comment by MaartenGr** [comment]
Not necessarily, Sentence Transformers takes in the document as a whole and then embeds the text. Splitting text up into sentences is also quite difficult as you cannot simply separate sentences by a dot when it could happen within a sentence. 

If documents are not too long (within the token limit) or are generally about the same topic throughout the document, there should be no problem using it for semantic similarity within BERTopic. However, when documents are starting to become not only a bit too long but also contain multiple topics, then it is highly advised to either split the text up into paragraphs or into sentences. Both would work well and splitting up into paragraphs decreases the number of data points, which is what I would typically advise.

**Comment by nilsblessing** [comment]
@pankratz-l In case you want to go with sentence-level embeddings, you could try out a sentencizer which is for example included in the SpaCy library. It is able to identify sentences based on a language model rather than based on punctuations only. However, depending on the data, you have to be careful, because you might lose context by splitting sentences. In such a case, the splitting by paragraphs mentioned by @MaartenGr might be more appropriate.

Here is an example code snippet for the SpaCy Sentencizer:

```
import spacy

nlp = spacy.load("en_core_web_sm")
doc = nlp("This is an example... This is another one.")
sent_list = [sent.text for sent in doc.sents]
```

**Comment by pankratz-l** [comment]
Thank you! I will try out splitting on \n and with the Sentencizer and see if the results improve.

**Comment by aph61** [comment]
@pankratz-l In most projects I worked on I considered the question "what is the unit of message". 
My documents had a natural structure of document - paragraphs (text between <p></p>) - sentences, where sentences were pragraph parts separated by ".", "?", "!", and ";". The question is, what is the natural unit of information; a sentence within the paragraph (the other sentences are "glue" to the message) or "paragraph" (the other paragraphs are glue). Turns out that the "unit of information" is easy to find: "glue" will have topic -1. In that case the number of topic=-1 sentences will be huge. In my case (characterize the content of the web page) I used the paragraph unit of information. 30% of the paragraphs had topic=-1, but taking this over the full document (several paragraphs) only 3% of the documents had no identifiable content. (Turns out, these pages had too much identifiable content, for example an overview of movies with a short description. Deeper analysis (one movie at a time) gave excellent results)

So in my case (characterize complete web pages) taking paragraphs as unit of information worked best. I did not care too much about the grammar, just about the semantics
